Template Updates

cratelyn opened the pull request:
## Template Changes

Here are some small changes to our template program. These are intended to address a few corners I've seen new users running into, namely: setting the correct backend name, and how users should handle errors in our application.

### Overview

The most important changes are...

* Bump `fastly` to version `0.3.0`
* Use the new `#[fastly::main]` procedural macro. (See: https://docs.rs/fastly/0.3.0/fastly/attr.main.html)
* This moves our `"backend_name"` literals up to the top of the file, into documented `const` values. This helps highlight the parts of our template that a user will always need to change, so people do not forget and run into cryptic hostcall error messages.
* This also adds a small comment to the `main` documentation explaining that returning an error will result in a 500 Internal Server Error response being sent back to the client. I think this is a small conceptual detail that isn't explained clearly currently.

#### Small Changes

A few other small changes made in the process of this PR:
* Adds documentation comments to elements.
* Moves the `VALID_METHODS` array into our `handle_request` function. This was done to try and emphasize the `BACKEND_NAME` consts, which **must** be changed for someone's service to work.

Commit:
template updates [..]

This makes a few minor stylistic changes to our `main.rs` template:

* use builder to generate internal error response
* describe error case in doc comment
* move `main` to top of program
* move valid methods array into `handle_request`
* move backend constants to top, add docs

cratelyn commented on src/main.rs:
```diff
@@ -1,21 +1,62 @@
 use std::convert::TryFrom;
 
-const VALID_METHODS: [Method; 3] = [Method::HEAD, Method::GET, Method::POST];
+/// The name of a backend server associated with this service.
+///
+/// This should be changed to match the name of your own backend.
+const BACKEND_NAME: &str = "backend_name";
```
I would really like a link to some sort of `docs.fastly` page somewhere that explains how/where to configure backends in our UI or via the CLI, but I expect C@E is young enough that this doesn't exist yet. Lmk if you have any ideas!

As mentioned in the description, I put this in a const above the rest of the code as a way to quietly highlight that this because it **must** be changed.

triblondon commented on src/main.rs:
```diff
@@ -1,21 +1,63 @@
+/// [dsreq]: ../fastly/request/fn.downstream_request.html
+/// [err]: https://tools.ietf.org/html/rfc7231#section-6.6.1
+/// [resp]: ../fastly/struct.Response.html
+/// [send]: ../fastly/response/trait.ResponseExt.html#method.send_downstream
```
Not super keen on this markdowny syntax - does this output nice docs somewhere?  My feeling is that the customer is likely to be interacting with this content *directly*, rather than viewing a rendered version of it somewhere else - so it may be better to refer to syntax more literally.  Otherwise as a total rust noob I might be tempted to write something like

```
let req = [`fastly::downstream_request`][dsreq];
```

softprops replied:
Rustdoc, rust's documentation generating backend, is a explicitly a markdown renderer. This style of linking docs is pretty common in the rust ecosystem.  

https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#basic-structure 

Rustdoc is built into the rust toolchain. A common workflow for browsing documentation for a repo is running `cargo doc --open` to generate and open docs in your local browser.

Rust customers will be familiar with this. Customers new to rust might benefit from a readme note on rust workflows.

cratelyn replied:
Thanks for pointing this out @triblondon, these are really good points! Folks are almost always going to be interacting with these comments directly as source. While addressing this, I realized something else: we don't export `main`, so the output in `cargo doc` won't include our functions anyhow. Ex:

![default-program-docs](https://user-images.githubusercontent.com/57912822/79788871-50204500-8317-11ea-8657-bea92d239fb1.png)

So, fancy comments really don't do much good here. Removed!

triblondon commented on src/main.rs:
```diff
@@ -1,21 +1,63 @@
+/// [send]: ../fastly/response/trait.ResponseExt.html#method.send_downstream
+fn main() -> Result<(), Error> {
+    let req = downstream_request()?;
+    let resp = match handle_request(req) {
```
A comment on this line might be nice - pointing the user to the function that follows.

cratelyn replied:
Good eye! This ends up being outdated via `8176c42`. Let me know what you think!

triblondon reviewed (commented):
Left a few comments.  I do think the new comment format makes it less clear.  No real opinion on the order of the functions.  Moving backend names into constants is a good idea.

Commit:
address @triblondon pr review

This PR removes the links in the documentation comments of `main`
and `handle_request` in our template.

While Rust documentation comments often include markdown links, this
program template actually isn't a fantastic place to use this
functionality.

Most often, these comments will be viewed directly in the source code.

Another important point against these is that neither of these functions
are exported with `pub`. Consequentially, they won't show up in `cargo
doc --open`. At best, they would appear in auto-completion dialogues
in editors like VSCode, but this benefit isn't worth the impact to
readability.

Commit:
update `fastly` to 0.3.0

Commit:
add `#[fastly::main]` to template

cratelyn commented on Cargo.toml:
```diff
@@ -8,4 +8,4 @@ edition = "2018"
 
 [dependencies]
-fastly = "^0.2.0-alpha4"
+fastly = "^0.3.0"
```
We should bump the `fastly` version we're using here.

cratelyn commented on src/main.rs:
```diff
@@ -1,21 +1,34 @@
-/// synthetic responses.
-fn handle_request(mut req: Request<Body>) -> Result<Response<Body>, Error> {
+/// If `main` returns an error a 500 error response will be delivered to the client.
+#[fastly::main]
+fn main(mut req: Request<Body>) -> Result<impl ResponseExt, Error> {
```
:tada: 

We have a `fastly::main` macro now, which helps remove a lot of boilerplate :tada: 

https://docs.rs/fastly/0.3.0/fastly/attr.main.html

cratelyn replied:
I've updated the documentation for this function, pulling the relevant bits of `main` and `handle_request`. Happy to bikeshed :bike:

cratelyn commented on src/main.rs:
```diff
@@ -51,21 +64,3 @@ fn handle_request(mut req: Request<Body>) -> Result<Response<Body>, Error> {
-/// should ultimately call `send_downstream` on a fastly::Response to deliver an
-/// HTTP response to the client.
-fn main() -> Result<(), Error> {
-    let req = downstream_request()?;
-    match handle_request(req) {
-        Ok(resp) => resp.send_downstream()?,
```
`#[fastly::main]` can do all of this for us, so `fn main` is gone now.

phamann reviewed (approved):
:+1: LGTM

acfoltzer reviewed (approved):
Looks great!
